### PR TITLE
Landan/deeplabv3 ade20k example

### PR DESCRIPTION
**Pull request by Landanjs**
# What does this PR do?

Contributes example script to train DeepLabv3+ on ADE20k.

Old WandB runs for the full 128 epochs and without any constant LR:

|           | mIoU | Time-to-Train (hours) | 
| ----- |  ----- | ----- |
| [Baseline + dice](https://wandb.ai/mosaic-ml/landan-test/runs/3gz09yh1) | 46.3 | 3.85 | 
| [Mild](https://wandb.ai/mosaic-ml/landan-test/runs/ze8v8qhg) | 46.84 | 3.31 | 
| [Medium](https://wandb.ai/mosaic-ml/landan-test/runs/2w4gouvh) | 49.13 | 5.0 |
| [Hot](https://wandb.ai/mosaic-ml/landan-test/runs/3mx17bbq) | 48.22 | 6.68 | 

I plan to rerun with constant LR and the target number of epochs for each recipe

### CODE REVIEW

File: examples/segmentation/train_deeplabv3_ade20k.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+    # Weight and Biases logger if specified in commandline
+    logger = None
+    if args.wandb_logger:
+        logging.info('Building Weights and Biases logger')
+        if args.wandb_entity is None:
+            raise ValueError('Please specify --wandb_entity argument')
+        if args.wandb_project is None:
+            raise ValueError('Please specify --wandb_project argument')
+        if args.wandb_run_name is None:
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply examples/segmentation/train_deeplabv3_ade20k.py:325]: I forgot about that, I'll delete it, thanks!

**Comment by Landanjs** [comment]
Updated description with previous WandB runs, will attempt new runs with constant LR and appropriate recipe number of epochs

File: docs/source/conf.py [hunk 2]
```diff
@@ -139,7 +139,9 @@ def _get_commit_sha() -> str:
 # List of patterns, relative to source directory, that match files and
 # directories to ignore when looking for source files.
 # This pattern also affects html_static_path and html_extra_path.
-exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store', 'examples/imagenet/README.md']
+exclude_patterns = [
+    '_build', 'Thumbs.db', '.DS_Store', 'examples/imagenet/README.md', 'examples/segmentation/README.md'
```
Comment (Landanjs) [inline docs/source/conf.py:143]: @eracah is it okay for me to ignore the README.md here? the bash section in the readme leads to issues with doctests which I wasn't sure how to resolve

File: examples/segmentation/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+# Mild DeepLabv3+ recipe for fastest training to 45.6 mIoU:
+composer train_deeplabv3_ade20k.py /path/to/ade20k/ --recipe_name mild --max_duration 25ep
+
+# Medium DeepLabv3+ recipe for highest mIoU (49.15) with similar training time as baseline:
+composer train_deeplabv3_ade20k.py /path/to/ade20k/ --recipe_name medium --max_duration 90ep
+
+# Hot DeepLabv3+ recipe for highest mIoU (49.83) with a long training schedule:
+composer train_deeplabv3_ade20k.py /path/to/ade20k --recipe_name hot --max_duration 256ep
+```
```
Comment (Landanjs) [inline examples/segmentation/README.md:36]: TODO: need to update these....

File: examples/segmentation/train_deeplabv3_ade20k.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+parser.add_argument('--run_name', help='Name of the training run used for checkpointing and logging', type=str)
+parser.add_argument('--seed', help='Random seed', type=int, default=17)
+parser.add_argument('--max_duration',
+                    help='Duration to train specified as a Time string',
+                    type=Time.from_timestring,
+                    default='128ep')
+
+args = parser.parse_args()
+
+IMAGENET_CHANNEL_MEAN = (int(0.485 * 255), int(0.456 * 255), int(0.406 * 255))
```
Comment (coryMosaicML) [inline examples/segmentation/train_deeplabv3_ade20k.py:97]: Should be ADE20k mean, std here right?

File: examples/segmentation/train_deeplabv3_ade20k.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+                         split='validation',
+                         both_transforms=None,
+                         image_transforms=image_transforms,
+                         target_transforms=target_transforms)
+
+    #Create ADE20k validation dataloader
+
+    val_sampler = None
+    if dist.get_world_size():
+        # Nifty function to instantiate a PyTorch DistributedSAmpler based on your hardware
```
Comment (coryMosaicML) [inline examples/segmentation/train_deeplabv3_ade20k.py:206]: `PyTorch DistributedSAmpler` -> `PyTorch DistributedSampler`

File: examples/segmentation/train_deeplabv3_ade20k.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        algorithms = [
+            ChannelsLast(),
+            EMA(half_life='1000ba', update_interval='10ba'),
+            SAM(rho=0.3, interval=2),
+            MixUp(alpha=0.2),
+        ]
+    elif args.recipe_name == 'hot':
+        algorithms = [
+            ChannelsLast(),
+            EMA(half_life='2000ba', update_interval='10ba'),
```
Comment (coryMosaicML) [inline examples/segmentation/train_deeplabv3_ade20k.py:311]: Hot recipe uses `update_interval='1ba'` for EMA.

File: examples/segmentation/README.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# Semantic Segmentation Example
+
+This example illustrates how to train a semantic segmentation model in composer
+
```
Comment (coryMosaicML) [inline examples/segmentation/README.md:4]: Maybe include a bit of installation instructions in the readme? Or, mention a docker image one can use.

**Review by coryMosaicML — APPROVED** [review]
Great work! A few minor comments.

Biggest things:
Maybe include a little info on how to install things, or which docker image to use
One of the params for the hot recipe was a bit off.

File: examples/segmentation/train_deeplabv3_ade20k.py (hunk 4 above)
Comment (Landanjs) [reply examples/segmentation/train_deeplabv3_ade20k.py:97]: We use the mean and standard deviation from ImageNet, I'm guessing I got this from MMseg and the rational was that it is the normalization used for the pretrained model?

File: examples/segmentation/train_deeplabv3_ade20k.py (hunk 5 above)
Comment (Landanjs) [reply examples/segmentation/train_deeplabv3_ade20k.py:206]: Ooops, fixed!

File: examples/segmentation/train_deeplabv3_ade20k.py (hunk 6 above)
Comment (Landanjs) [reply examples/segmentation/train_deeplabv3_ade20k.py:311]: Ooops, fixed!

File: examples/segmentation/README.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+## DeepLabv3+ on ADE20k
+
+The `train_deeplabv3_ade20k.py` script trains a DeepLabv3+ model with either a ResNet-50 or ResNet-101 on the ADE20k semantic segmentation benchmark.
+
+Before running the script, ADE20k semantic segmentation benchmark must be downloaded from [here](http://data.csail.mit.edu/places/ADEchallenge/ADEChallengeData2016.zip), then unzip. The path to the dataset will be used as the first argument to the script.
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply examples/segmentation/README.md:9]: Ooops, I'll delete this and just mention the option to automatically download

**Review by knighton — APPROVED** [review]
